In [34]:
# Dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [35]:
# Load csv into Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [36]:
# Summary stats
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [37]:
# Plot the DF
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

### Preparing Data

In [38]:
# Normalize the data from the CSV file
market_data_scaled = StandardScaler().fit_transform(
    df_market_data[['price_change_percentage_24h', 'price_change_percentage_7d', 
                    'price_change_percentage_14d', 'price_change_percentage_30d', 
                    'price_change_percentage_60d', 'price_change_percentage_200d', 
                    'price_change_percentage_1y']]
)

In [39]:
# DataFrame with the scaled data
df_market_scaled = pd.DataFrame(market_data_scaled, 
                                columns= ['price_change_percentage_24h', 'price_change_percentage_7d', 
                                          'price_change_percentage_14d', 'price_change_percentage_30d', 
                                          'price_change_percentage_60d', 'price_change_percentage_200d', 
                                          'price_change_percentage_1y'])

# Set the coin_id as index
df_market_scaled['coin_id'] = df_market_data.index
df_market_scaled = df_market_scaled.set_index('coin_id')

# Preview df
df_market_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

### Find the Best Value for k Using the Original Data.

In [40]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))

In [41]:
inertia = []
for i in k:
    model = KMeans(n_clusters=i, random_state=0)
    model.fit(df_market_scaled)
    inertia.append(model.inertia_)

In [42]:
# Create a dictionary with the data to plot the Elbow curve
elbow_curve_data = {'k': k, 'inertia': inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_curve_data)

# Review df
df_elbow.head()

k     inertia
0  1  287.000000
1  2  195.820218
2  3  123.190482
3  4   79.022435
4  5   65.302379

In [43]:
# Identify the optimal value for k using a line chart
elbow_chart = df_elbow.hvplot.line(x='k', y= 'inertia', title= 'Elbow Curve', xticks=k)
elbow_chart

:Curve   [k]   (inertia)

The best value for k is 3.

### Cluster Cryptocurrencies with K-means Using the Original Data

In [44]:
# Initialize the K-Means model
model = KMeans(n_clusters= 3, random_state=0)

In [45]:
# Fit using scaled data
model.fit(df_market_scaled)

KMeans(n_clusters=3, random_state=0)

In [46]:
# Predict the clusters to group the cryptocurrencies using the scaled data
market_clusters = model.predict(df_market_scaled)

# Print the resulting array of cluster values.
print(market_clusters)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 2
 0 0 0 0]


In [47]:
# Copy df
df_market_predictions = df_market_scaled.copy()

In [48]:
# DataFrame with predicted clusters
df_market_predictions["market_cluster"] = market_clusters

# Preview df
df_market_predictions.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  market_cluster  
coin_id                                                   
bitcoin                        -0.251637               0  
ethereum                       -0.199352               0  
tether                         -0.282061               0  
ripple                         -0.295546               0  
bitcoin-cash                   -0.270317               0

In [49]:
# Cluster scatterplot
df_market_predictions.hvplot.scatter(
    x = 'price_change_percentage_24h',
    y = 'price_change_percentage_7d',
    by = 'market_cluster',
    hover_cols = ['coin_id'],
    title = 'Scatter Plot by Segment - k=4'
)

:NdOverlay   [market_cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

### Optimize Clusters with Principal Component Analysis.

In [50]:
# PCA model instance n_components=3
pca = PCA(n_components=3)


In [51]:
# Reduce to three principal components
market_pca_data = pca.fit_transform(df_market_scaled)

# Review df
market_pca_data[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [52]:
# Retrieve the explained variance to determine how much information can be attributed to each principal component
explained_variance_ratios = pca.explained_variance_ratio_
explained_variance_ratios

array([0.3719856 , 0.34700813, 0.17603793])

In [53]:
total_explained_variance = sum(explained_variance_ratios)
total_explained_variance

0.8950316570309841

**Question:** What is the total explained variance of the three principal components?

**Answer:** Total explained variance is just the sum of all three principal components. In this case: 0.8950316570309841

In [54]:
# Create a new DataFrame with the PCA data
df_market_pca = pd.DataFrame(market_pca_data, columns= ['PC1', 'PC2', 'PC3'])

# Copy the crypto names from the original data
df_market_pca['coin_id'] = df_market_data.index

# Set the coin_id as index
df_market_pca = df_market_pca.set_index('coin_id')

# Review df
df_market_pca.head()

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

### Find the Best Value for k Using the PCA Data

In [55]:
# Create a list with the number of k-values from 1 to 11
k_pca = list(range(1, 11))

In [56]:
# Create an empy list to store the inertia values
inertia_pca = []

# Create a for loop to compute the inertia with each possible value of k
for i_pca in k_pca:
    model = KMeans(n_clusters= i_pca, random_state=0)
    model.fit(df_market_pca)
    inertia_pca.append(model.inertia_)

In [57]:
# Create a dictionary with the data to plot the Elbow curve
elbow_curve_pca = {'k': k_pca, 'inertia': inertia_pca}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca = pd.DataFrame(elbow_curve_pca)

# Review df
df_elbow_pca.head()

k     inertia
0  1  256.874086
1  2  165.901994
2  3   93.774626
3  4   49.665497
4  5   37.878747

In [58]:
# Identify the optimal value for k using a line chart
elbow_chart_pca = df_elbow_pca.hvplot.line(x='k', y= 'inertia', title= 'Elbow Curve', xticks=k_pca)
elbow_chart_pca

:Curve   [k]   (inertia)

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best value for k is 4

* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** Yes, the best value for k using the original data was 3

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [59]:
# Initialize the K-Means mode using the best value for k
model = KMeans(n_clusters=4)

In [60]:
# Fit the K-Means model using the PCA data
model.fit(df_market_pca)

KMeans(n_clusters=4)

In [61]:
# Predict the clusters to group the cryptocurrencies using the PCA data
market_clusters_pca = model.predict(df_market_pca)

# Print the resulting array of cluster values.
print(market_clusters_pca)

[3 3 0 0 3 3 3 3 3 0 0 0 0 3 0 3 0 0 3 0 0 3 0 0 0 0 0 0 3 0 0 0 1 3 0 0 2
 0 0 0 0]


In [65]:
# Copy df
df_market_pca_predictions = df_market_pca.copy()

# DataFrame with predicted clusters
df_market_pca_predictions["market_cluster"] = market_clusters_pca

# Preview df
df_market_pca_predictions.head()

PC1       PC2       PC3  market_cluster
coin_id                                                   
bitcoin      -0.600667  0.842760  0.461595               3
ethereum     -0.458261  0.458466  0.952877               3
tether       -0.433070 -0.168126 -0.641752               0
ripple       -0.471835 -0.222660 -0.479053               0
bitcoin-cash -1.157800  2.041209  1.859715               3

In [66]:
# Cluster scatterplot
df_market_pca_predictions.hvplot.scatter(
    x = 'PC1',
    y = 'PC2',
    by = 'market_cluster',
    hover_cols = ['coin_id'],
    title = 'Scatter Plot by Segment - k=4'
)

:NdOverlay   [market_cluster]
   :Scatter   [PC1]   (PC2,coin_id)